test/extensions/filters/network/common/fuzz/uber_per_readfilter.cc

```diff
@@ -151,6 +151,26 @@ void UberFilterFuzzer::checkInvalidInputForFuzzer(const std::string& filter_name
           "http_conn_manager trying to use Quiche which we won't fuzz here. Config:\n{}",
           config.DebugString()));
     }
+    if (config.codec_type() == envoy::extensions::filters::network::http_connection_manager::v3::
+                                   HttpConnectionManager::HTTP2) {
+      // Sanity check on connection_keepalive interval and timeout.
+      try {
+        PROTOBUF_GET_MS_REQUIRED(config.http2_protocol_options().connection_keepalive(), interval);
+      } catch (const DurationUtil::OutOfRangeException& e) {
+        throw EnvoyException(
+            absl::StrCat("In http2_protocol_options.connection_keepalive interval shall not be "
+                         "negative. Exception {}",
+                         e.what()));
+      }
+      try {
+        PROTOBUF_GET_MS_REQUIRED(config.http2_protocol_options().connection_keepalive(), timeout);
+      } catch (const DurationUtil::OutOfRangeException& e) {
+        throw EnvoyException(
+            absl::StrCat("In http2_protocol_options.connection_keepalive timeout shall not be "
+                         "negative. Exception {}",
+                         e.what()));
+      }
```

Comment:
I think if we just do validation of the proto we should get this. 

See the pgv annotation: https://github.com/envoyproxy/envoy/blob/5cc7d31ba986b0046364772057f7fa8e6301e0bc/api/envoy/config/core/v3/protocol.proto#L336-L342

Reply:
I was also under the same assumption, but it turns out that if Nanos is negative and seconds is positive then pgv doesn't pick it up.
See thread in private repo

Reply:
:( no access to private repo thread, but ultimately is the duration > 1000000ns ? if so that seems valid still since we will be using the duration not the individual components.

Reply:
This is an excerpt from the private repo:
 
After digging a bit it seems that this is the offending proto:
```
http2_protocol_options {
  connection_keepalive {
    interval {
      seconds: 150323855360
      nanos: 134217728
    }
    timeout {
      seconds: 150323855360
      nanos: -100663296
    }
  }
}
```

The `timeout` field does have a [PGV constraint](https://github.com/envoyproxy/envoy/blob/7eb3a87b8757e030aedfdc4959adc509e89ac788/api/envoy/config/core/v3/protocol.proto#L340) so it must be >= 1000000, and IIUC if one "merges" the seconds and nanos together the timeout it is indeed more than the threshold.
Envoy has a [stricter validation](https://github.com/envoyproxy/envoy/blob/0e8899c90213b39d8f4a1a083d4fd31e9c2fe8c1/source/common/protobuf/utility.cc#L912) and requires both to be >= 0.

@htuch also noted that:
> We have to validate this in code, rather than proto constraints, since we can't use PGV in the WKTs (and also PGV might not be expressive enough).
>
> The right place to verify is within the code backing macros like PROTOBUF_GET_MS_REQUIRED.

Reply:
And furthermore is the `Duration` proto defined to have `sign(seconds) == sign(nanos)`, which can not be checked by PGV constraints and therefore is done in code.

Any further input required?

Reply:
makes sense, thanks for clarifying.